python/dgl/nn/pytorch/conv/pnaconv.py

```diff
@@ -0,0 +1,255 @@
+    >>>
+    >>> g = dgl.graph(([0,1,2,3,2,5], [1,2,3,4,0,3]))
+    >>> feat = th.ones(6, 10)
+    >>> conv = PNAConv(10, 10, ('mean', 'max', 'sum'), ('identity', 'amplification'), 2.5)
+    >>> ret = conv(g, feat)
+    """
+    def __init__(self, in_size, out_size, aggregators, scalers, delta, dropout=0., num_towers=1, edge_feat_size=0):
+        super(PNAConv, self).__init__()
+        aggregators = [AGGREGATORS[aggr] for aggr in aggregators]
+        scalers = [SCALERS[scale] for scale in scalers]
+        
+        self.in_size = in_size
+        self.out_size = out_size
+        self.tower_in_size = in_size // num_towers
+        self.tower_out_size = out_size // num_towers
+        self.edge_feat_size = edge_feat_size
+        
+        self.towers = nn.ModuleList([
+            PNAConvTower(
+                self.tower_in_size, self.tower_out_size,
+                aggregators, scalers, delta,
+                dropout=dropout, edge_feat_size=edge_feat_size
+            ) for _ in range(num_towers)
+        ])
+        
+        self.mixing_layer = nn.Sequential(
+            nn.Linear(out_size, out_size),
+            nn.LeakyReLU()
+        )
+        
+    def forward(self, graph, node_feat, snorm_n=None, edge_feat=None):
+        r"""
+        Description
+        -----------
+        Compute PNA layer.
+
+        Parameters
+        ----------
+        graph : DGLGraph
+            The graph.
+        node_feat : torch.Tensor
+            The input feature of shape :math:`(N, h_n)`. :math:`N` is the number of
+            nodes, and :math:`h_n` must be the same as in_size.
+        snorm_n : torch.Tensor, optional
+            The normalization tensor of shape :math:`(N, 1)` related to the size of
+            the graph, computed as :math:`[[|g_1|^{-1}], [|g_1|^{-1}], [|g_2|^{-1}],\cdots]`.
+            If not provided, we calculate it according to the overall size of the input graph.
+            The parameter is required for a batched graph.
+        edge_feat : torch.Tensor, optional
+            The edge feature of shape :math:`(M, h)`. :math:`M` is the number of
+            edges, and :math:`h_e` must be the same as edge_feat_size.
+
+        Returns
+        -------
+        torch.Tensor
+            The output node feature of shape :math:`(N, h_n')` where :math:`h_n'`
+            should be the same as out_size.
+        """
+        N = graph.num_nodes()
+        if snorm_n is None:
+            snorm_n = (torch.ones(N, 1).to(node_feat) / N).sqrt()
+        h_cat = torch.cat([
+            tower(
+                g,
+                node_feat[:, ti * self.tower_in_size: (ti + 1) * self.tower_in_size],
+                snorm_n,
+                edge_feat
+            )
+            for ti, tower in enumerate(self.towers)
+        ], dim=1)
+        h_out = self.mixing_layer(h_cat)
+        # add residual connection
+        if self.in_size == self.out_size:
+            h_out = h_out + node_feat
```

Comment:
Is this the original implementation in PNA? I prefer having an explicit input argument for residual connection like `res=True`. If `in_size` is not equal to `out_size`, then use a linear layer to project input node features.